### PR TITLE
Marketplace contract

### PR DESCRIPTION
**Pull request by geofmureithi**
This introduces a marketplace contracts.
#81

### CODE REVIEW

File: contracts/marketplace/src/lib.rs
Comment (mariopil) [inline contracts/marketplace/src/lib.rs:122]: I would add an event here informing about asset being sold.

File: contracts/marketplace/src/lib.rs
Comment (mariopil) [inline contracts/marketplace/src/lib.rs:159]: Also event missing here.

File: contracts/marketplace/src/lib.rs
Comment (mariopil) [inline contracts/marketplace/src/lib.rs:146]: Why `assert_eq` is used here and the prices below are compared using `if` and `panic_with_error`?

File: contracts/marketplace/src/lib.rs
Comment (mariopil) [inline contracts/marketplace/src/lib.rs:188]: Event...

File: contracts/marketplace/src/lib.rs
Comment (mariopil) [inline contracts/marketplace/src/lib.rs:179]: Also here, once `assert_eq` is used and then `if` + `panic_with_error `

File: contracts/marketplace/src/lib.rs
Comment (mariopil) [inline contracts/marketplace/src/lib.rs:208]: Here too...

File: contracts/marketplace/src/lib.rs
Comment (mariopil) [inline contracts/marketplace/src/lib.rs:211]: Event...

File: contracts/marketplace/Makefile.toml
Comment (mariopil) [inline contracts/marketplace/Makefile.toml:113]: Wrong input args

File: contracts/marketplace/Makefile.toml
Comment (mariopil) [inline contracts/marketplace/Makefile.toml:123]: Not existing function, I guess it was not updated yet?

File: contracts/marketplace/Makefile.toml
Comment (mariopil) [inline contracts/marketplace/Makefile.toml:134]: Same here...

File: contracts/marketplace/Makefile.toml
Comment (mariopil) [inline contracts/marketplace/Makefile.toml:166]: In the future we should think about moving some of the common tasks into separate file. Practically every test will call `calculate_fees` and probably few more.

**Review by mariopil — CHANGES_REQUESTED** [review]
I'm missing events. Let say a user is interested in one asset and would like to be updated whenever its price changes or it is sold.

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs:179]: One just panics if you are not the owner.
The other returns an error that your input is wrong.
All these just do the same thing only that `panic_with_error` adds some specific error.

Comment (mariopil) [reply contracts/marketplace/src/lib.rs:179]: Maybe better add an error saying user is not an owner? It would be more clear what is wrong.

**Review by mariopil — APPROVED** [review]

File: contracts/marketplace/src/lib.rs
Comment (eloylp) [inline contracts/marketplace/src/lib.rs:76]: Is this covered in tests ?

File: contracts/marketplace/src/lib.rs
Comment (eloylp) [inline contracts/marketplace/src/lib.rs:101]: Is this covered in tests ?

File: contracts/marketplace/src/lib.rs
Comment (eloylp) [inline contracts/marketplace/src/lib.rs:251]: Is this covered in tests ?

File: contracts/marketplace/src/lib.rs
Comment (eloylp) [inline contracts/marketplace/src/lib.rs:44]: Should we check in all other contract functions that this method was invoked (initilized) ?

File: contracts/marketplace/src/lib.rs
Comment (eloylp) [inline contracts/marketplace/src/lib.rs:180]: Is this covered in tests ?

File: contracts/marketplace/src/lib.rs
Comment (eloylp) [inline contracts/marketplace/src/lib.rs:60]: This is stored, but when its used ? Is it the idea to retain a sales fee for the the transaction for the admin of the contract ?

File: contracts/marketplace/Makefile.toml
Comment (eloylp) [inline contracts/marketplace/Makefile.toml:3]: Should we start using `testnet` instead of futurenet ? 

Looks like futurenet could be disabled at some point, now that we are in testnet.

**Review by eloylp — COMMENTED** [review]
I think this is looking good. I left a couple of comments to understand certain elements. Also some comments regarding test coverage of the invariants.

Great job so far !

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs:81]: I messaged you on slack. This does not work. I have no idea why, was looking for pointers.

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs:60]: Yes, the admin would set on init.

Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs:60]: Hmm, it should during buy listing

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs:44]: But it would just panic right?

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs:49]: True, Unit would be better here

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs:58]: I am actually wondering if we should remove the part for percentages.
Leave that to anyone working reusing the contract.

Comment (eloylp) [reply contracts/marketplace/src/lib.rs:62]: Not sure. I would involve here @mariopil also. 

My thoughts: 

If this wasnt a smart contract, i would say to implement it. Theres no marketplace out there that doesnt charge fees on their global sales. But taken into account this is a smart contract, things could bedifferent.

If we are in doubt we can always add it. Then set it to zero depending on the use case.

File: contracts/marketplace/src/test.rs
Comment (eloylp) [inline contracts/marketplace/src/test.rs:65]: Should this `asset` be registered with `create_token_asset` instead of registering `token` ? On my mind `token` just represent the token with which we are going to trade things. Could this be part of the error we are getting [here](https://github.com/eigerco/nebula/pull/82#discussion_r1343736870) @geofmureithi ?

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs]: Done

File: contracts/marketplace/src/lib.rs
Comment (geofmureithi) [reply contracts/marketplace/src/lib.rs]: Created issue #90

File: contracts/marketplace/src/test.rs
Comment (geofmureithi) [reply contracts/marketplace/src/test.rs:45]: This is referred to issue #90

Comment (eloylp) [reply contracts/marketplace/src/test.rs:45]: Ok, i understand we are going to look at this separatelly/inmediatelly after this merge right ? Because under my understanding, currently there is no real owner change more than in our contract internal state.

**Review by eloylp — APPROVED** [review]

**Pull request merged**
